python/ray/serve/long_poll.py

```diff
@@ -37,52 +48,69 @@ class LongPollAsyncClient:
-        assert in_async_loop, "The client is only available in async context."
-        asyncio.get_event_loop().create_task(self._do_long_poll())
+    def update_key_listeners(
+            self,
+            key_listeners: Dict[KeyType, UpdateStateCallable],
+    ):
+        self.key_lisners = key_listeners
+        self._reset()
 
     def _poll_once(self) -> ray.ObjectRef:
-        object_ref = self.host_actor.listen_for_change.remote(
+        self._current_ref = self.host_actor.listen_for_change.remote(
             self.snapshot_ids)
-        return object_ref
-
-    async def _do_long_poll(self):
-        while True:
-            try:
-                updates: Dict[str, UpdatedObject] = await self._poll_once()
-                logger.debug("LongPollClient received updates for keys: "
-                             f"{list(updates.keys())}.")
-                for key, update in updates.items():
-                    self.object_snapshots[key] = update.object_snapshot
-                    self.snapshot_ids[key] = update.snapshot_id
-                    # NOTE(simon):
-                    # This blocks the loop from doing another poll. Consider
-                    # use loop.create_task here or poll first then call the
-                    # callbacks.
-                    callback = self.key_listeners[key]
-                    await callback(update.object_snapshot)
-            except ray.exceptions.RayActorError:
-                # This can happen during shutdown where the controller is
-                # intentionally killed, the client should just gracefully
-                # exit.
-                logger.debug("LongPollClient failed to connect to host. "
-                             "Shutting down.")
-                break
+        self._current_ref._on_completed(
+            lambda update: self._process_update(update))
+
+    def _process_update(self, updates: Dict[str, UpdatedObject]):
+        if isinstance(updates, (ray.exceptions.RayActorError)):
+            # This can happen during shutdown where the controller is
+            # intentionally killed, the client should just gracefully
+            # exit.
+            logger.debug("LongPollClient failed to connect to host. "
+                         "Shutting down.")
+            return
+
+        if isinstance(updates, (ray.exceptions.RayTaskError)):
+            # This can happen during shutdown where the controller doesn't
+            # contain this key, we will just repull.
+            # NOTE(simon): should we repull or just wait in the long poll
+            # host?
+            if not isinstance(updates.as_instanceof_cause(), ValueError):
+                logger.error("LongPollHost errored\n" + updates.traceback_str)
+            self._poll_once()
+            return
+
+        # Before we process the updates and calling callbacks, kick off
+        # another poll so we can pipeline the polling and processing.
+        self._poll_once()
```

Comment:
This behavior seems a little bit weird - this will cause it to essentially poll repeatedly forever. I'd expect if I call `_poll_once` that it polls only once :)

Reply:
Can we handle this at a higher layer? Or should we make this `_poll_forever()`?